### PR TITLE
Storage: Create snapshot mount path before mounting

### LINKED ISSUE #7124: lxd snap v4.0.0: lxc commands hang for some time
This is on an Ubuntu 18.04 host running kernel 4.15.0-91-generic

Since the automatic update of the lxd snap to v4.0.0 basic lxc commands hang for almost a minute,e.g.:
```
$ time lxc list
[...]
real  0m51,010s
user  0m0,084s
sys   0m0,103s
```

`lxc info CONTAINER` shows output until here:
```
[...]
  Network usage:
    eth1:
      Bytes received: 9.77MB
      Bytes sent: 375.59MB
      Packets received: 94351
      Packets sent: 41851
    lo:
      Bytes received: 48.59MB
      Bytes sent: 48.59MB
      Packets received: 168219
      Packets sent: 168219
```
then hangs for the same amount of time before finally showing the snapshots.

This is with few containers (around 10). Storage backend is ZFS. Reboot of the server did not help. I'm happy to provide more info, just tell me where to look. I didn't see anything strange in the logs so far.

Here's some output from `strace lxc info CONTAINER`:
```
[...]
getsockname(6, {sa_family=AF_UNIX}, [112->2]) = 0
getpeername(6, {sa_family=AF_UNIX, sun_path="/var/snap/lxd/common/lxd/unix.socket"}, [112->39]) = 0
futex(0xc000100148, FUTEX_WAKE_PRIVATE, 1) = 1
write(6, "GET /1.0 HTTP/1.1\r\nHost: unix.so"..., 144) = 144
futex(0xc000180148, FUTEX_WAKE_PRIVATE, 1) = 1
epoll_pwait(3, [], 128, 0, NULL, 824635513768) = 0
futex(0x1383a28, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
epoll_pwait(3, [], 128, 0, NULL, 0)     = 0
epoll_pwait(3, [{EPOLLOUT, {u32=212315976, u64=140308203941704}}], 128, -1, NULL, 0) = 1
epoll_pwait(3, [], 128, 0, NULL, 20633544) = 0
futex(0x1383a28, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
epoll_pwait(3, [], 128, 0, NULL, 0)     = 0
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f9c0c840000
openat(AT_FDCWD, "/etc//localtime", O_RDONLY) = 6
read(6, "TZif2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\5\0\0\0\5\0\0\0\0"..., 4096) = 1918
read(6, "", 4096)                       = 0
close(6)                                = 0
futex(0xc000180148, FUTEX_WAKE_PRIVATE, 1) = 1
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 6
setsockopt(6, SOL_SOCKET, SO_BROADCAST, [1], 4) = 0
connect(6, {sa_family=AF_UNIX, sun_path="/var/snap/lxd/common/lxd/unix.socket"}, 39) = 0
epoll_ctl(3, EPOLL_CTL_ADD, 6, {EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, {u32=212315976, u64=140308203941704}}) = 0
getsockname(6, {sa_family=AF_UNIX}, [112->2]) = 0
getpeername(6, {sa_family=AF_UNIX, sun_path="/var/snap/lxd/common/lxd/unix.socket"}, [112->39]) = 0
futex(0xc000180148, FUTEX_WAKE_PRIVATE, 1) = 1
read(6, 0xc000476000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
futex(0x1383a28, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x1383a28, FUTEX_WAIT_PRIVATE, 0, NULL
```
and this is where it hangs. Does that help?

PS: just occurred to me, should I have filed this under lxc/lxc?

### PR DESCRIPTION
**Pull request by tomponline**
Fixes #7124

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>

### CODE REVIEW

**Comment by tomponline** [comment]
Im aware of the test fails on here

**Comment by tomponline** [comment]
@stgraber ready for review.

File: lxd/storage/drivers/volume.go [hunk 1]
```diff
@@ -145,23 +153,36 @@ func (v Volume) EnsureMountPath() error {
...
+	compareMode := os.ModeDir | mode
+
+	// Set mode of actual volume's mount path if needed.
+	if fInfo.Mode() != compareMode {
+		err = os.Chmod(volPath, mode)
+
+		// If the chmod failed, return the error as long as the volume is not a snapshot.
+		// If the volume is a snapshot, we must ignore the error as snapshots are readonly and cannot be
+		// modified after they are taken, such that any permission error is not fixable at mount time.
+		if err != nil && !v.IsSnapshot() {
```
Comment (tomponline) [inline lxd/storage/drivers/volume.go:171]: @stgraber in the end I decided to just skip chmod errors for snapshots, as any error that occurs here is not fixable, and we should just continue as best we can. For non-snapshot vols, any error here is critical and must return an error.

Comment (stgraber) [reply lxd/storage/drivers/volume.go:171]: Sounds good

**Pull request merged**